<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 15px; border-radius: 10px; text-align: center; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
  Project Title: <span style="color: #333333;">Sales Insights For Atliq Hardware</span>
</h1>

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 20px; border-radius: 15px; text-align: center; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
  <span style="color: #1c1c1c; font-size: 2rem;">👨‍💻</span> 
  Author: <span style="color: #333333;">Maaz Umar</span><br>
  <a href="https://github.com/Maaz-Umar-00" style="color: #333333; text-decoration: none; font-size: 0.9rem;">GitHub</a> |
  <a href="https://www.linkedin.com/in/maaz-umar-178090307/" style="color: #333333; text-decoration: none; font-size: 0.9rem;">LinkedIn</a> |
  <a href="mailto:maazumar2005@email.com" style="color: #555555; text-decoration: none; font-size: 0.9rem;">Email</a>

</h1>

<img src="./atliq_hardware_company_sales_insights_written_on_dashboard.jpeg" alt="Mitron Bank" style="width: 600px; display: block; margin: 0 auto;" />

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 15px; border-radius: 10px; text-align: center; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
  Meta-Data(About Dataset) <span style="color: #333333;"></span>
</h1>

This dataset belongs to **Atliq Hardware** and contains sales and transaction data, which needs to be analyzed to get sales insights. The dataset consists of 5 tables with the following columns:

1. **customers**
2. **date**
3. **markets**
4. **products**
5. **transactions**

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
Table Descriptions and Column Details <span style="color: #333333;"></span>
</h1>

### Table 1: `customers`
- **customer_code**: Unique code for each customer.
- **customer_name**: Name of the customer.
- **customer_type**: Type of customer (e.g., Brick).

### Table 2: `date`
- **order_date**: The date when an order was placed.
`It contains other date related columns`

### Table 3: `markets`
- **market_code**: Unique code for each market.
- **market_name**: Name of the market.
- **zone**: The geographic zone or region of the market.

### Table 4: `products`
- **product_code**: Unique code for each product.
- **product_type**: Type of product.

### Table 5: `transactions`
- **product_code**: Code of the product involved in the transaction.
- **customer_code**: Code of the customer who made the purchase.
- **market_code**: Code of the market where the transaction occurred.
- **order_date**: Date when the transaction took place.
- **sales_qty**: Quantity of the product sold.
- **currency**: The currency used in the transaction.
- **profit_margin_percentage**: Percentage of profit margin for the transaction.
- **profit_margin**: The absolute profit made in the transaction.
- **cost_price**: The cost price of the product sold.


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
Project Aim: <span style="color: #333333;"></span>
</h1>

- The goal of this project is to analyze the sales data of **Atliq Hardware** to gain valuable insights into sales performance, customer behavior, market trends, and profitability. By exploring this data, we aim to identify key patterns, optimize sales strategies, and improve decision-making for the business.

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
Key Findings:<span style="color: #333333;"></span>
</h1>

- **Atliq Hardware's revenue peaked at $43 million in 2018, but declined by approximately 65% to $15 million by 2020.**
- **Atliq Hardware** could consider implementing a subscription model for regular customers, allowing them to receive monthly supplies at a discounted rate, thereby ensuring steady revenue. 
- **Atliq Hardware** receives the highest revenue from **Delhi NCR**, where profits are also higher; however, the **profit margin** is surprisingly very low. 

To address this issue, the company should analyze cost structures in the region to identify inefficiencies and explore opportunities for reducing operational costs. Additionally, focusing on premium product offerings could enhance profit margins while attracting discerning customers.
- **The highest profit margin** for Atliq Hardware is from **Surat city**, showcasing strong profitability. However, in contrast, the profit margin in **Bengaluru** is declining and moving into the **negative.** 

To mitigate the declining profits in **Bengaluru,** the company should reassess its pricing strategy and consider reducing non-essential expenditures in that region. Additionally, conducting a market analysis to understand **changing customer preferences** could help identify new opportunities for growth.


- The company sells  the most product to these top 5 stores:
  - **Electrical Sara Stores**
  - **Electricalsylitical**
  - **Excel Stores**
  - **Premium Stores**
  - **Nixon**


[Click here to Download  the Power Bi Dashboard](https://github.com/Maaz-Umar-00/Sales-Insights-For-Atliq-hardware/blob/main/02_Atliq_hardware_sales_project.pbix)
<hr style="border-top: 2px solid #16A085;">


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 15px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
<span style="color: #333333;">1.Importing Libraries:</span>
</h1> 

In [1]:
# importing libraries
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
import mysql
import mysql.connector
import os 


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
<span style="color: #333333;">1.1 Connecting to the database</span>
</h1> 

In [2]:
df=mysql.connector.connect(host='localhost',
                           username='root',
                           password='#####',
                           database='sales')
cursor=df.cursor()

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
<span style="color: #333333;">1.2 Taking Basic Overveiw of tansactions Tables</span>
</h1>


In [ ]:
# Executing the query
query = """ SELECT * FROM transactions; """
cursor.execute(query)

# Fetching all results
results = cursor.fetchall()

# Geting column names from cursor.description
column_names = [desc[0] for desc in cursor.description]

print("Columns Names Are",column_names)

# printing the 5 rows 
for row in results[:5]:
    print(row)

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
<span style="color: #333333;">1.3 checking for total number of columns and rows in transactions table
</span>
</h1>


In [7]:
cursor.execute(f"SELECT COUNT(*) FROM transactions")
total_rows = cursor.fetchone()[0]

cursor.execute(f"SHOW COLUMNS FROM transactions")
columns_info = cursor.fetchall()
total_columns = len(columns_info)

# Output the results
print(f"Total number of rows in transactions table are: {total_rows}")
print(f"Total number of  Columns in transactions table are: {total_columns}")

Total number of rows in transactions table are: 148395
Total number of  Columns in transactions table are: 10


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
<span style="color: #333333;">1.4 Taking Basic overveiw of  customers table</span>
</h1>


In [ ]:
query = """ SELECT * FROM customers; """
cursor.execute(query)

# Fetching all results
results = cursor.fetchall()

# Geting column names from cursor.description
column_names = [desc[0] for desc in cursor.description]

print("Columns Names Are",column_names)

# printing the 5 rows 
for row in results[:5]:
    print(row)

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
<span style="color: #333333;">1.5 checking for total number of columns and rows in customers table
</span>
</h1>


In [8]:
cursor.execute(f"SELECT COUNT(*) FROM customers")
total_rows = cursor.fetchone()[0]

cursor.execute(f"SHOW COLUMNS FROM customers")
columns_info = cursor.fetchall()
total_columns = len(columns_info)

# Output the results
print(f"Total number of rows in customers table are: {total_rows}")
print(f"Total number of  Columns in customers table are: {total_columns}")

Total number of rows in customers table are: 38
Total number of  Columns in customers table are: 3


### Observations
---
- **Transactions Table:** 148,395 rows and 10 columns.
- **Customers Table:** 38 rows and 3 columns.
- We will primarily work with these tables as they contain many useful columns for our analysis.


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 15px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.5rem;">
2. Before going on further analysis, let's Check Data is accurate or not..<span style="color: #333333;">
</span>
</h1>


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
<span style="color: #333333;">2.1 First Checking for duplicates in both tables.
</span>
</h1>


In [5]:
duplicates_query_transactions = """
SELECT *, COUNT(*) as count
FROM transactions
GROUP BY product_code, customer_code, market_code, order_date, sales_qty, sales_amount, currency, profit_margin_percentage, profit_margin, cost_price
HAVING count > 1;
"""
cursor.execute(duplicates_query_transactions)
duplicates_transactions = cursor.fetchall()
# Output the results
if duplicates_transactions:
    print("Duplicates in Transactions Table:")
    for row in duplicates_transactions:
        print(row)
else:
    print("Luckily, we didn't find any duplicates in the Transactions Table.")


Luckily, we didn't find any duplicates in the Transactions Table.


In [9]:
duplicates_query_customers = """
SELECT *, COUNT(*) as count
FROM customers
GROUP BY customer_code, custmer_name, customer_type
HAVING count > 1;
"""

# Execute the query
cursor.execute(duplicates_query_customers)
duplicates_customers = cursor.fetchall()


if duplicates_customers:
    print("\nDuplicates in Customers Table:")
    for row in duplicates_customers:
        print(row)
else:
    print("Luckily, we didn't find any duplicates in the Customers Table.")

Luckily, we didn't find any duplicates in the Customers Table.


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
<span style="color: #333333;">2.2 Now check for missing values  in both Tables.
</span>
</h1>


In [10]:
missing_values_query_transactions = """
SELECT 
    COUNT(*) - COUNT(product_code) AS product_code_nulls,
    COUNT(*) - COUNT(customer_code) AS customer_code_nulls,
    COUNT(*) - COUNT(market_code) AS market_code_nulls,
    COUNT(*) - COUNT(order_date) AS order_date_nulls,
    COUNT(*) - COUNT(sales_qty) AS sales_qty_nulls,
    COUNT(*) - COUNT(sales_amount) AS sales_amount_nulls,
    COUNT(*) - COUNT(currency) AS currency_nulls,
    COUNT(*) - COUNT(profit_margin_percentage) AS profit_margin_percentage_nulls,
    COUNT(*) - COUNT(profit_margin) AS profit_margin_nulls,
    COUNT(*) - COUNT(cost_price) AS cost_price_nulls
FROM transactions;
"""

cursor.execute(missing_values_query_transactions)
missing_transactions = cursor.fetchall()

if all(value == 0 for value in missing_transactions[0]):
    print("We didn't find any missing values in the Transactions table.")
else:
    print("Missing Values in Transactions Table:")
    print(missing_transactions)


We didn't find any missing values in the Transactions table.


In [12]:
missing_values_query_customers = """
SELECT 
    COUNT(*) - COUNT(customer_code) AS customer_code_nulls,
    COUNT(*) - COUNT(custmer_name) AS customer_name_nulls,
    COUNT(*) - COUNT(customer_type) AS customer_type_nulls
FROM customers;
"""
cursor.execute(missing_values_query_customers)
missing_customers = cursor.fetchall()

# Check if there are any missing values in the customers table
if all(value == 0 for value in missing_customers[0]):
    print("We didn't find any missing values in the Customers table.")
else:
    print("Missing Values in Customers Table:")
    print(missing_customers)

We didn't find any missing values in the Customers table.


**Luckily we didn't have duplicates and missing values in both tables and now let's move on to check for inconsistences or inaccurate data in the table.**


<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 12px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1.2rem;">
<span style="color: #333333;">2.3 Let's check for inconsistent or inaccurate data in both tables.
</span>
</h1>


In [21]:
column_names = [desc[0] for desc in cursor.description]
print(column_names)

['product_code', 'customer_code', 'market_code', 'order_date', 'sales_qty', 'sales_amount', 'currency', 'profit_margin_percentage', 'profit_margin', 'cost_price']


In [ ]:
query = """
SELECT * from transactions
"""
[desc[0] for desc in cursor.description]
# printing the 5 rows 
for row in results:
    print(row)

<h1 style="font-family: 'Poppins', sans-serif; font-weight: 700; color: #1c1c1c; background: white; padding: 8px; border-radius: 10px; text-align: left; box-shadow: 0 4px 8px rgba(0, 0, 0, 0.5); letter-spacing: 1px; text-transform: capitalize; font-size: 1rem;">
<span style="color: #333333;">2.3.1 checking unique values in the currency column
</span>
</h1>


In [27]:
query = """
SELECT DISTINCT(currency) FROM transactions;
"""

# Execute the query
cursor.execute(query)

# Fetch and print the unique currencies
unique_currencies = cursor.fetchall()
for currency in unique_currencies:
    print(currency)

('INR',)
('USD',)


### Observations
-----
- I am noticing that there is a `-1` in the sales amount, which is not possible because the sales amount cannot be negative.
- Additionally, the currency column contains both `INR` and `USD`, but since this data is primarily in `INR`, we will convert the `USD` values to `INR`.


  Now, we will move on to Power BI, where we will perform data cleaning and analysis.

<h2 style="color: #D35400; font-family: Arial, sans-serif;">Now, We will move on to <span style="font-weight: bold; color: #16A085;">Power Bi</span> where we will perform data cleaning and analysis.</h2>
<p style="font-size: 18px; color: #2C3E50; font-family: 'Georgia', serif;">
Power Bi provides advanced visualizations and allows us to explore the data more effectively. Stay tuned for insights!
</p>

[Click here to download Power Bi Dashboard](https://github.com/Maaz-Umar-00/Sales-Insights-For-Atliq-hardware/blob/main/02_Atliq_hardware_sales_project.pbix)
<hr style="border-top: 2px solid #16A085;">

`